File: src/main/java/com/tfmg_energy_connector/ConverterBlockEntity.java
package com.tfmg_energy_connector;

import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.world.level.block.entity.BlockEntity;
import net.minecraft.world.level.block.state.BlockState;
import net.neoforged.neoforge.capabilities.Capabilities;
import net.neoforged.neoforge.energy.EnergyStorage;
import net.neoforged.neoforge.energy.IEnergyStorage;

public class ConverterBlockEntity extends BlockEntity {

    // Хранилище: 100,000 емкость, 10,000 вход/выход
    private final EnergyStorage energyStorage = new EnergyStorage(100000, 10000, 10000) {
        @Override
        public int receiveEnergy(int maxReceive, boolean simulate) {
            setChanged(); // Сохраняем блок при изменении энергии
            return super.receiveEnergy(maxReceive, simulate);
        }

        @Override
        public int extractEnergy(int maxExtract, boolean simulate) {
            setChanged();
            return super.extractEnergy(maxExtract, simulate);
        }
    };

    public ConverterBlockEntity(BlockPos pos, BlockState blockState) {
        super(TfmgAe2Bridge.CONVERTER_BE.get(), pos, blockState);
    }

    // Этот метод вызывается 20 раз в секунду
    public void tick() {
        if (level == null || level.isClientSide) return;

        // Если есть энергия, пытаемся отдать её соседям
        if (energyStorage.getEnergyStored() > 0) {
            pushEnergyToNeighbors();
        }
    }

    private void pushEnergyToNeighbors() {
        for (Direction direction : Direction.values()) {
            if (energyStorage.getEnergyStored() <= 0) return;

            BlockPos neighborPos = worldPosition.relative(direction);
            // Получаем энергию соседа через новую систему NeoForge
            IEnergyStorage neighborEnergy = level.getCapability(Capabilities.EnergyStorage.BLOCK, neighborPos, direction.getOpposite());

            if (neighborEnergy != null && neighborEnergy.canReceive()) {
                int toSend = energyStorage.extractEnergy(1000, true); // Проверяем, сколько можем вынуть
                int accepted = neighborEnergy.receiveEnergy(toSend, false); // Отдаем
                energyStorage.extractEnergy(accepted, false); // Вынимаем по-настоящему
            }
        }
    }

    // Сохранение при выходе из игры
    @Override
    protected void saveAdditional(CompoundTag tag, net.minecraft.core.HolderLookup.Provider registries) {
        super.saveAdditional(tag, registries);
        tag.putInt("Energy", energyStorage.getEnergyStored());
    }

    // Загрузка при входе
    @Override
    protected void loadAdditional(CompoundTag tag, net.minecraft.core.HolderLookup.Provider registries) {
        super.loadAdditional(tag, registries);
        int energy = tag.getInt("Energy");
        // Сброс и установка (хак для EnergyStorage)
        energyStorage.extractEnergy(Integer.MAX_VALUE, false);
        energyStorage.receiveEnergy(energy, false);
    }

    // Геттер для регистрации Capability
    public IEnergyStorage getEnergyStorage() {
        return this.energyStorage;
    }
}

File: src/main/java/com/tfmg_energy_connector/ConverterBlockEntity.kt
package com.tfmg_energy_connector

import com.drmangotea.tfmg.content.electricity.base.IElectric
import com.drmangotea.tfmg.content.electricity.base.ElectricBlockValues
import com.drmangotea.tfmg.content.electricity.base.ElectricalNetwork
import net.minecraft.core.BlockPos
import net.minecraft.core.Direction
import net.minecraft.core.HolderLookup
import net.minecraft.nbt.CompoundTag
import net.minecraft.world.level.LevelAccessor
import net.minecraft.world.level.block.entity.BlockEntity
import net.minecraft.world.level.block.state.BlockState
import net.neoforged.neoforge.capabilities.Capabilities
import net.neoforged.neoforge.energy.EnergyStorage

class ConverterBlockEntity(pos: BlockPos, state: BlockState) :
    BlockEntity(TfmgAe2Bridge.CONVERTER_BE_TYPE.get(), pos, state), IElectric {


    val energyStorage = object : EnergyStorage(100, 100, 100) {
        override fun canReceive() = true
        override fun canExtract() = true
    }


    private val electricValues = ElectricBlockValues(pos.asLong())

    override fun getBlockPos(): BlockPos {
        return worldPosition
    }

    override fun getPos(): Long {
        return worldPosition.asLong()
    }


    override fun getData(): ElectricBlockValues = electricValues

    override fun getLevelAccessor(): LevelAccessor = level!!

    override fun sendStuff() {
        setChanged()
        level?.sendBlockUpdated(worldPosition, blockState, blockState, 3)
    }

    override fun setNetwork(networkId: Long) {
        data.electricalNetworkId = networkId
    }

    override fun resistance(): Float = 20f

    override fun getPowerUsage(): Int {
        val space = energyStorage.maxEnergyStored - energyStorage.energyStored
        return if (space > 0) minOf(space, 1024) else 0
    }

    override fun voltageGeneration(): Int = 0
    override fun powerGeneration(): Int = 0

    override fun onNetworkChanged(oldVoltage: Int, oldPower: Int) {
        setChanged()
    }

    override fun setVoltage(v: Int) {
        data.voltage = v
    }

    override fun setNetworkResistance(r: Float) {
        data.networkResistance = r.toInt()
    }

    fun tick() {
        val lvl = level ?: return
        if (lvl.isClientSide || isRemoved) return

        tickElectricity()
        if (lvl.gameTime % 20 == 0L) {
            lazyTickElectricity()
        }

        if (data.voltage > 0 && !data.notEnoughPower) {
            val powerToTake = getPowerUsage()
            if (powerToTake > 0) {
                energyStorage.receiveEnergy(powerToTake, false)

            }
        }

        if (energyStorage.energyStored > 0) {
            for (dir in Direction.entries) {
                val cap = lvl.getCapability(Capabilities.EnergyStorage.BLOCK, worldPosition.relative(dir), dir.opposite)
                cap?.let {
                    if (it.canReceive()) {
                        val push = energyStorage.extractEnergy(100, true)
                        val accepted = it.receiveEnergy(push, false)
                        energyStorage.extractEnergy(accepted, false)
                    }
                }
            }
        }
    }


    override fun saveAdditional(tag: CompoundTag, registries: HolderLookup.Provider) {
        super.saveAdditional(tag, registries)
        tag.putInt("fe_stored", energyStorage.energyStored)
    }

    override fun loadAdditional(tag: CompoundTag, registries: HolderLookup.Provider) {
        super.loadAdditional(tag, registries)
        energyStorage.receiveEnergy(tag.getInt("fe_stored"), false)
    }
}